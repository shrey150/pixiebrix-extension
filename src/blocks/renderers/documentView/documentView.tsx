/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React from "react";
import { Card, Col, Container, Row } from "react-bootstrap";
import { get } from "lodash";
import { UnknownObject } from "@/types";
import { isPipelineExpression } from "@/runtime/mapArgs";
import PipelineComponent from "./PipelineComponent";
import DocumentButton from "./DocumentButton";

const headerComponents = {
  header_1: "h1",
  header_2: "h2",
  header_3: "h3",
} as const;

const gridComponents = {
  container: Container,
  row: Row,
  column: Col,
} as const;

const UnknownType: React.FC<{ componentType: string }> = ({
  componentType,
}) => <span>Unknown type: {componentType}</span>;

export function getComponent(
  body: any
): { Component: React.ComponentType | string; props: UnknownObject } {
  const componentType = String(body.type);
  const config = get(body, "config", {});

  switch (componentType) {
    case "header_1":
    case "header_2":
    case "header_3": {
      const { title, ...props } = config;
      props.children = title;

      // eslint-disable-next-line security/detect-object-injection -- componentType is header_1, header_2, or header_3
      return { Component: headerComponents[componentType], props };
    }

    case "text": {
      const { text, ...props } = config;
      props.children = text;

      return { Component: "p", props };
    }

    case "container":
    case "row":
    case "column": {
      const props = { ...config };
      props.children = body.children.map((child: any, i: number) => {
        const { Component: ChildComponent, props: childProps } = getComponent(
          child
        );
        return <ChildComponent key={i} {...childProps} />;
      });

      // eslint-disable-next-line security/detect-object-injection -- componentType is container, row, or column
      return { Component: gridComponents[componentType], props };
    }

    case "card": {
      const { heading, body, ...cardProps } = config;
      const Component = () => (
        <Card {...cardProps}>
          <Card.Header>{heading}</Card.Header>
          <Card.Body>
            {isPipelineExpression(body) ? (
              <PipelineComponent pipeline={body.__value__} />
            ) : (
              body
            )}
          </Card.Body>
        </Card>
      );
      return {
        Component,
        props: undefined,
      };
    }

    case "block": {
      const pipeline = get(body, "config.pipeline");
      if (!isPipelineExpression(pipeline)) {
        throw new Error("Expected pipeline expression for pipeline");
      }

      return {
        Component: PipelineComponent,
        props: {
          pipeline: pipeline.__value__,
        },
      };
    }

    case "button": {
      const { title, onClick, ...props } = config;
      if (!isPipelineExpression(onClick)) {
        throw new Error("Expected pipeline expression for onClick");
      }

      return {
        Component: DocumentButton,
        props: {
          children: title,
          onClick: onClick.__value__,
          ...props,
        },
      };
    }

    default: {
      return {
        Component: UnknownType,
        props: { componentType: componentType ?? "No Type Provided" },
      };
    }
  }
}
