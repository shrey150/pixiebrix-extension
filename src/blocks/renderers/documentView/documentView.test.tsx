/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from "@testing-library/react";
import React from "react";
import { getComponent } from "./documentView";

const renderDocument = (config: any) => {
  const { Component, props } = getComponent(config);
  return render(<Component {...props} />);
};

test.each(["header_1", "header_2", "header_3"])(
  "renders header, %s",
  (headerType: string) => {
    const config = {
      type: headerType,
      config: {
        title: "Test Header",
        className: "test-class",
      },
    };
    const rendered = renderDocument(config);
    expect(rendered.asFragment()).toMatchSnapshot();
  }
);

test("renders paragraph text", () => {
  const config = {
    type: "text",
    config: {
      text: "Test Paragraph",
      className: "test-class",
    },
  };
  const rendered = renderDocument(config);
  expect(rendered.asFragment()).toMatchSnapshot();
});

test("renders unknown type", () => {
  const config = {
    type: "TheTypeForWhichAComponentIsNotDefined",
    className: "test-class",
  };
  const rendered = renderDocument(config);
  expect(rendered.asFragment()).toMatchSnapshot();
});

test("renders grid", () => {
  const config = {
    type: "container",
    config: {
      className: "text-primary",
    },
    children: [
      {
        type: "row",
        children: [
          {
            type: "column",
            children: [
              {
                type: "header_1",
                config: {
                  title: "Header",
                },
              },
            ],
          },
        ],
      },
      {
        type: "row",
        config: {
          className: "mt-5",
        },
        children: [
          {
            type: "column",
            config: {
              className: "p-3",
            },
            children: [
              {
                type: "text",
                config: {
                  text: "left column",
                },
              },
            ],
          },
          {
            type: "column",
            config: {
              className: "p-5",
            },
            children: [
              {
                type: "text",
                config: {
                  text: "right column",
                },
              },
            ],
          },
        ],
      },
    ],
  };

  const rendered = renderDocument(config);
  expect(rendered.asFragment()).toMatchSnapshot();
});

test("renders button", () => {
  const config = {
    type: "button",
    config: {
      title: "Button under test",
      variant: "primary",
      onClick: {
        __type__: "pipeline",
        __value__: jest.fn(),
      },
    },
  };
  const rendered = renderDocument(config);
  expect(rendered.asFragment()).toMatchSnapshot();
});
