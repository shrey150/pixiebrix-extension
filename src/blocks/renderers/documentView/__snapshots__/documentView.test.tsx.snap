// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card renders pipeline body 1`] = `
<DocumentFragment>
  <div
    class="test-class card"
  >
    <div
      class="card-header"
    >
      Test Heading of Card
    </div>
    <div
      class="card-body"
    >
      <div
        class="text-danger p-3"
      >
        Error running panel: No handler for ECHO_SENDER was registered in the receiving end
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`card renders text body 1`] = `
<DocumentFragment>
  <div
    class="test-class card"
  >
    <div
      class="card-header"
    >
      Test Heading of Card
    </div>
    <div
      class="card-body"
    >
      Test body of card
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders button 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary"
    type="button"
  >
    Button under test
  </button>
</DocumentFragment>
`;

exports[`renders grid 1`] = `
<DocumentFragment>
  <div
    class="text-primary container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          Header
        </h1>
      </div>
    </div>
    <div
      class="mt-5 row"
    >
      <div
        class="p-3 col"
      >
        <p>
          left column
        </p>
      </div>
      <div
        class="p-5 col"
      >
        <p>
          right column
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders header, header_1 1`] = `
<DocumentFragment>
  <h1
    class="test-class"
  >
    Test Header
  </h1>
</DocumentFragment>
`;

exports[`renders header, header_2 1`] = `
<DocumentFragment>
  <h2
    class="test-class"
  >
    Test Header
  </h2>
</DocumentFragment>
`;

exports[`renders header, header_3 1`] = `
<DocumentFragment>
  <h3
    class="test-class"
  >
    Test Header
  </h3>
</DocumentFragment>
`;

exports[`renders paragraph text 1`] = `
<DocumentFragment>
  <p
    class="test-class"
  >
    Test Paragraph
  </p>
</DocumentFragment>
`;

exports[`renders unknown type 1`] = `
<DocumentFragment>
  <span>
    Unknown type: TheTypeForWhichAComponentIsNotDefined
  </span>
</DocumentFragment>
`;
